File: tech-dhtml/revisarCodigo/game-connect4-main/src/models/Board.js
import { Coordinate } from '../types/Coordinate.js'
import { Color } from '../types/Color.js'
import { Line } from './Line.js'
import { Direction } from '../types/Direction.js'
import {assert} from "../utils/assert.js";

export class Board {
  #colors
  #lastDrop

  constructor() {
    this.#colors = []
    for (let i = 0; i < Coordinate.NUMBER_ROWS; i++) {
      this.#colors[i] = []
      for (let j = 0; j < Coordinate.NUMBER_COLUMNS; j++) {
        this.#colors[i][j] = Color.NULL
      }
    }
  }

  dropToken(column, color) {
    assert(!this.isComplete(column))
    assert(!this.isWinner())
    this.#lastDrop = new Coordinate(0, column)
    while (!this.isEmpty(this.#lastDrop)) {
      this.#lastDrop = this.#lastDrop.shifted(Direction.NORTH.getCoordinate())
    }
    this.#colors[this.#lastDrop.getRow()][this.#lastDrop.getColumn()] = color

  }

  isComplete(column) {
    if (column !== undefined) {
      return !this.isEmpty(new Coordinate(Coordinate.NUMBER_ROWS - 1, column))
    }
    for (let i = 0; i < Coordinate.NUMBER_COLUMNS; i++) {
      if (!this.isComplete(i)) {
        return false
      }
    }
    return true
  }

  isFinished() {
    return this.isComplete() || this.isWinner()
  }

  isWinner() {
    if (this.#lastDrop === undefined) {
      return false
    }
    let line = new Line(this.#lastDrop)
    for (let direction of Direction.values().splice(0, 4)) {
      line.set(direction)
      for (let i = 0; i < Line.LENGTH; i++) {
        if (this.isConnect4(line)) {
          return true
        }
        line.shift()
      }
    }
    return false
  }

  isConnect4(line) {
    let coordinates = line.getCoordinates()
    for (let i = 0; i < Line.LENGTH; i++) {
      if (!coordinates[i].isValid()) {
        return false
      }
      if (
        i > 0 &&
        this.getColor(coordinates[i - 1]) !== this.getColor(coordinates[i])
      ) {
        return false
      }
    }
    return true
  }

  isOccupied(coordinate, color) {
    return this.getColor(coordinate) === color
  }

  isEmpty(coordinate) {
    return this.isOccupied(coordinate, Color.NULL)
  }

  getColor(coordinate) {
    return this.#colors[coordinate.getRow()][coordinate.getColumn()]
  }

  getLastDrop() {
    return this.#lastDrop
  }
}

File: tech-java-orientadoObjetos/src/main/es/labturing/connect4/models/Board.java
package main.es.labturing.connect4.models;

import java.io.Serializable;

import main.es.labturing.connect4.types.Color;

public class Board implements Serializable {

    public static final int LINE_LENGTH = 4;
    protected Color[][] colors;
    protected Coordinate lastDrop;

    public Board() {
        this.colors = new Color[Coordinate.NUMBER_ROWS][Coordinate.NUMBER_COLUMNS];
        this.reset();
    }

    public void reset() {
        for (int i = 0; i < Coordinate.NUMBER_ROWS; i++) {
            for (int j = 0; j < Coordinate.NUMBER_COLUMNS; j++) {
                this.colors[i][j] = Color.NULL;
            }
        }
    }

    public void dropToken(int column, Color color) {
        this.lastDrop = new Coordinate(0, column);
        while (!this.isEmptyPosition(this.lastDrop)) {
            this.lastDrop = this.lastDrop.shifted(Direction.NORTH.getCoordinate());
        }
        this.colors[this.lastDrop.getRow()][this.lastDrop.getColumn()] = color;
    }

    public boolean isEmptyPosition(Coordinate coordinate) {
        return this.isOccupied(coordinate, Color.NULL);
    }

    private boolean isOccupied(Coordinate coordinate, Color color) {
        return this.getColor(coordinate) == color;
    }

    private boolean isComplete() {
        for (int i = 0; i < Coordinate.NUMBER_COLUMNS; i++) {
            if (!this.isCompleteColumn(i)) {
                return false;
            }
        }
        return true;
    }

    public boolean isCompleteColumn(int column) {
        return !this.isEmptyPosition(new Coordinate(Coordinate.NUMBER_ROWS - 1, column));
    }

    public boolean isGameFinished() {
        return this.isComplete() || this.isWinner();
    }

    public boolean isWinner() {
        if (this.lastDrop == null) {
            return false;
        }
        for (Direction direction : Direction.halfValues()) {
            Line line = new Line(this.lastDrop, direction, Board.LINE_LENGTH);
            for (int i = 0; i < Board.LINE_LENGTH; i++) {
                if (this.isConnect4(line)) {
                    return true;
                }
                line.shift();
            }
        }
        return false;
    }

    private boolean isConnect4(Line line) {
        for (int i = 0; i < Board.LINE_LENGTH; i++) {
            if (!line.getCoordinate(i).isValid()) {
                return false;
            }
            if (i > 0 && this.getColor(line.getCoordinate(i - 1)) != this.getColor(line.getCoordinate(i))) {
                return false;
            }
        }
        return true;
    }

    public Color getColor(Coordinate coordinate) {
        return this.colors[coordinate.getRow()][coordinate.getColumn()];
    }

    public void setColor(Color color, Coordinate coordinate) {
        this.colors[coordinate.getRow()][coordinate.getColumn()] = color;
    }

    public Color[][] getColors() {
        return this.colors;
    }

    public Coordinate getLastDrop() {
        return this.lastDrop;
    }

    public BoardState getState() {
        return new BoardState(this.colors, this.lastDrop);
    }

    public void setState(BoardState boardState) {
        this.colors = boardState.getColors();
        this.lastDrop = boardState.getLastDrop();
    }

    public void setLastDrop(Coordinate coordinate) {
        this.lastDrop = coordinate;
    }
}
